### PR TITLE
[LLVM] Implement Bmi2.MultiplyNoFlags

### PR DESCRIPTION
**Pull request by EgorBo**
See https://github.com/mono/mono/pull/16919#issuecomment-534765820 (cc @imhameed). Now it's legal to unlock Bmi2 for all users (all methods are implemented).

```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static uint Test1(uint a, uint b)
{
    return Bmi2.MultiplyNoFlags(a, b);
}

[MethodImpl(MethodImplOptions.NoInlining)]
static uint Test2(uint a, uint b, uint* lower)
{
    return Bmi2.MultiplyNoFlags(a, b, lower);
}

[MethodImpl(MethodImplOptions.NoInlining)]
static ulong Test3(ulong a, ulong b)
{
    return Bmi2.X64.MultiplyNoFlags(a, b);
}

[MethodImpl(MethodImplOptions.NoInlining)]
static ulong Test4(ulong a, ulong b, ulong* lower)
{
    return Bmi2.X64.MultiplyNoFlags(a, b, lower);
}
```
ASM:
```asm
loWorld_Program_Test1__uint_uint_:
0000000000000000	movl	%edi, %ecx
0000000000000002	movl	%esi, %eax
0000000000000004	imulq	%rcx, %rax
0000000000000008	shrq	$0x20, %rax
000000000000000c	retq


loWorld_Program_Test2__uint_uint_uint__:
0000000000000000	movl	%edi, %ecx
0000000000000002	movl	%esi, %eax
0000000000000004	imulq	%rcx, %rax
0000000000000008	movl	%eax, (%rdx)
000000000000000a	shrq	$0x20, %rax
000000000000000e	retq


loWorld_Program_Test3__ulong_ulong_:
0000000000000000	movq	%rdi, %rdx
0000000000000003	mulxq	%rsi, %rcx, %rax
0000000000000008	retq


loWorld_Program_Test4__ulong_ulong_ulong__:
0000000000000000	movq	%rdx, %rcx
0000000000000003	movq	%rdi, %rdx
0000000000000006	mulxq	%rsi, %rdx, %rax
000000000000000b	movq	%rdx, (%rcx)
000000000000000e	retq
```
Tested different inputs - the results are the same as on CoreCLR.
No idea why LLVM doesn't want to use mulx for `uint` (only for `ulong`)

### CODE REVIEW

File: mono/mini/simd-intrinsics-netcore.c [hunk 1]
```diff
@@ -721,14 +718,27 @@ emit_x86_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature
...
-			//// ulong MultiplyNoFlags(ulong left, ulong right, ulong* low) => MultiplyNoFlags(left, right, low);
-			//// uint MultiplyNoFlags(uint left, uint right)
-			//// uint MultiplyNoFlags(uint left, uint right, uint* low)
-			//return NULL;
-		//case SN_ZeroHighBits:
+		case SN_MultiplyNoFlags:
+			if (fsig->param_count == 2) {
+				MONO_INST_NEW (cfg, ins, is_64bit ? OP_MULX_H64 : OP_MULX_H32);
+				ins->dreg = alloc_ireg (cfg);
+				ins->sreg1 = args [0]->dreg;
```
Comment (vargaz) [inline mono/mini/simd-intrinsics-netcore.c]: Needs to be alloc_lreg () for 64 bit.

File: mono/mini/simd-intrinsics-netcore.c [hunk 2]
```diff
@@ -721,14 +718,27 @@ emit_x86_intrinsics (MonoCompile *cfg, MonoMethod *cmethod, MonoMethodSignature
...
+				ins->sreg2 = args [1]->dreg;
+				ins->type = is_64bit ? STACK_I8 : STACK_I4;
+				MONO_ADD_INS (cfg->cbb, ins);
+			} else if (fsig->param_count == 3) {
+				MONO_INST_NEW (cfg, ins, is_64bit ? OP_MULX_HL64 : OP_MULX_HL32);
+				ins->dreg = alloc_ireg (cfg);
+				ins->sreg1 = args [0]->dreg;
```
Comment (vargaz) [inline mono/mini/simd-intrinsics-netcore.c]: Ditto.

File: mono/mini/simd-intrinsics-netcore.c (hunk 1 above)
Comment (EgorBo) [reply mono/mini/simd-intrinsics-netcore.c]: oops, thanks!

Comment (EgorBo) [reply mono/mini/simd-intrinsics-netcore.c]: I revised all methods (currently these intrinsics are emitted only for #if AMD64 so it was ok but I guess we can also enable them for x86 (once we have a test lane for it).

**Comment by EgorBo** [comment]
@imhameed 

> https://bugs.llvm.org/show_bug.cgi?id=34232

Ah, ok, thanks!

> The use of LLVMInt128Type on 32-bit is a little weird, but seems benign.

It uses `long long` [here](https://github.com/llvm/llvm-project/blob/release/6.x/clang/lib/Headers/bmi2intrin.h#L86) as well for 32 bit but the llvm optimizer shrinks it to 64 anyway (instcombine or BDCE)

**Comment by EgorBo** [comment]
@monojenkins build failed

**Comment by imhameed** [comment]
@monojenkins build failed

**Pull request merged**